pallets/dapp-staking-v3/src/types.rs

```diff
@@ -1649,87 +1648,96 @@ impl<NT: Get<u32>, T: TierSlotsFunc, P: Get<FixedU128>> TiersConfiguration<NT, T
+            FixedU128::from_rational(
                 (base_number_of_slots - new_number_of_slots).into(),
                 new_number_of_slots.into(),
-            );
-
-            let mut new_tier_thresholds = params.tier_thresholds.clone();
-            new_tier_thresholds
-                .iter_mut()
-                .for_each(|threshold| match threshold {
-                    TierThreshold::DynamicTvlAmount { amount, .. } => {
-                        *amount = amount
-                            .saturating_add(delta_threshold_increase.saturating_mul_int(*amount));
-                    }
-                    _ => (),
-                });
-
-            new_tier_thresholds
+            )
         };
 
+        let new_tier_thresholds: BoundedVec<Balance, NT> = params
+            .tier_thresholds
+            .clone()
+            .iter()
+            .map(|threshold| {
+                Self::adjust_threshold(
+                    threshold,
+                    delta_threshold,
+                    total_issuance,
+                    new_number_of_slots >= base_number_of_slots,
+                )
+            })
+            .collect::<Vec<_>>()
+            .try_into()
+            .unwrap_or_default();
+
         Self {
-            number_of_slots: new_number_of_slots,
             slots_per_tier: new_slots_per_tier,
             reward_portion: params.reward_portion.clone(),
             tier_thresholds: new_tier_thresholds,
             _phantom: Default::default(),
         }
     }
+
+    // Update tier thresholds.
+    // In case number of slots increase, we decrease thresholds required to enter the tier.
+    // In case number of slots decrease, we increase the threshold required to enter the tier.
+    //
+    // According to formula: %delta_threshold = (100% / (100% - delta_%_slots) - 1) * 100%
+    //
+    // where delta_%_slots is simply: (base_num_slots - new_num_slots) / base_num_slots
+    //
+    // `base_num_slots` is the number of slots at the base native currency price.
+    //
+    // When these entries are put into the threshold formula, we get:
+    // = 1 / ( 1 - (base_num_slots - new_num_slots) / base_num_slots ) - 1
+    // = 1 / ( new / base) - 1
+    // = base / new - 1
+    // = (base - new) / new
+    //
+    // This number can be negative. In order to keep all operations in unsigned integer domain,
+    // formulas are adjusted like:
+    //
+    // 1. Number of slots has increased, threshold is expected to decrease
+    // %delta_threshold = (new_num_slots - base_num_slots) / new_num_slots
+    // new_threshold = base_threshold * (1 - %delta_threshold)
+    //
+    // 2. Number of slots has decreased, threshold is expected to increase
+    // %delta_threshold = (base_num_slots - new_num_slots) / new_num_slots
+    // new_threshold = base_threshold * (1 + %delta_threshold)
+    //
+    fn adjust_threshold(
+        threshold: &TierThreshold,
```

Comment:
Suggestion if you keep this - move this under `impl TierThreshold` - the first param is already `&TierThreshold` so it only seems natural.